core/metacling/src/TCling.cxx

```diff
@@ -1112,6 +1112,95 @@ static std::string GetModuleNameAsString(clang::Module *M, const clang::Preproce
+#endif
+
+  // Load core modules
+  // This should be vector in order to be able to pass it to LoadModules
+  std::vector<std::string> CoreModules = {"ROOT_Foundation_C","ROOT_Config",
+    "ROOT_Foundation_Stage1_NoRTTI", "Core", "RIO"};
+
+  // FIXME: Reducing those will let us be less dependent on rootmap files
+  static constexpr std::array<const char*, 3> ExcludeModules =
+    { { "Rtools", "RSQLite", "RInterface"} };
+
+  LoadModules(CoreModules, clingInterp);
+
+  // Take this branch only from ROOT because we don't need to preload modules in rootcling
+  if (!IsFromRootCling()) {
+    // Dynamically get all the modules and load them if they are not in core modules
+    clang::CompilerInstance &CI = *clingInterp.getCI();
+    clang::ModuleMap &moduleMap = CI.getPreprocessor().getHeaderSearchInfo().getModuleMap();
+    clang::Preprocessor &PP = CI.getPreprocessor();
+    std::vector<std::string> ModulesPreloaded;
+    for (auto I = moduleMap.module_begin(), E = moduleMap.module_end(); I != E; ++I) {
+      clang::Module *M = I->second;
+      assert(M);
+
+      std::string ModuleName = GetModuleNameAsString(M, PP);
+      if (!ModuleName.empty() &&
+          std::find(CoreModules.begin(), CoreModules.end(), ModuleName) == CoreModules.end()
+          && std::find(ExcludeModules.begin(), ExcludeModules.end(), ModuleName) == ExcludeModules.end()) {
+        if (M->IsSystem && !M->IsMissingRequirement)
+          LoadModule(ModuleName, clingInterp);
+        else if (!M->IsSystem && !M->IsMissingRequirement)
+          ModulesPreloaded.push_back(ModuleName);
+      }
+    }
+    LoadModules(ModulesPreloaded, clingInterp);
+  }
+
+  // Check that the gROOT macro was exported by any core module.
+  assert(clingInterp.getMacro("gROOT") && "Couldn't load gROOT macro?");
+
+  // C99 decided that it's a very good idea to name a macro `I` (the letter I).
+  // This seems to screw up nearly all the template code out there as `I` is
+  // common template parameter name and iterator variable name.
+  // Let's follow the GCC recommendation and undefine `I` in case any of the
+  // core modules have defined it:
+  // https://www.gnu.org/software/libc/manual/html_node/Complex-Numbers.html
+  clingInterp.declare("#ifdef I\n #undef I\n #endif\n");
+
+  // These macros are from loading R related modules, which conflict with
+  // user's code.
+  clingInterp.declare("#ifdef PI\n #undef PI\n #endif\n");
+  clingInterp.declare("#ifdef ERROR\n #undef ERROR\n #endif\n");
+}
+
+static void RegisterPreIncludedHeaders(cling::Interpreter& clingInterp) {
+  std::string PreIncludes;
+  bool hasCxxModules = clingInterp.getCI()->getLangOpts().Modules;
+
+   // For the list to also include string, we have to include it now.
+   // rootcling does parts already if needed, e.g. genreflex does not want using
+   // namespace std.
+   if (IsFromRootCling()) {
+     PreIncludes += "#include \"RtypesCore.h\"\n";
+   } else {
+     if (!hasCxxModules)
```

Comment:
This part of the code seems to be 'new' (i.e. add support for hasCxxModules).    If this the case, it ought to be mentioned in the commit log (and the NFC should be removed).  Thanks.

Reply:
Yeah, ran clang-format.

Reply:
Thanks.  Can you also update the commit to mention the new functionality?

Reply:
Done.